pkgs/applications/virtualization/xen/4.18/4.18.nix

```diff
@@ -0,0 +1,194 @@
+  ovmfVersion = "202305";
+
+  src = fetchgit {
+    url = "git://xenbits.xenproject.org/xen.git";
+    rev = "RELEASE-${version}";
+    hash = "sha256-pHCjj+Bcy4xQfB9xHU9fccFwVdP2DXrUhdszwGvrdmY=";
+  };
+
+  # Sources needed to build tools and firmwares.
+  prefetchedSources =
+    lib.attrsets.optionalAttrs withInternalQEMU {
+      qemu-xen = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen.git";
+          rev = "qemu-xen-${qemuVersion}";
+          hash = "sha256-BX+LXfNzwdUMALwwI1ZDW12dJ357oynjnrboLHREDGQ=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace scripts/tracetool.py \
+            --replace-fail "/usr/bin/env python" "${python311}/bin/python"
+        '';
+        meta.description = "Xen's fork of upstream QEMU";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalTraditionalQEMU {
+      qemu-xen-traditional = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen-traditional.git";
+          rev = "xen-${qemuVersion}";
+          hash = "sha256-R+I8QUFTeWzc/EIHibxfCoZa/wzcR59izVF4ciVshrU=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace xen-hooks.mak \
+            --replace-fail /usr/include/pci ${pciutils}/include/pci
+        '';
+        meta.description = "Xen's fork of upstream QEMU that uses the traditional device model";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalSeaBIOS {
+      "firmware/seabios-dir-remote" = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/seabios.git";
+          rev = "rel-${seaBIOSVersion}";
+          hash = "sha256-hWemj83cxdY8p+Jhkh5GcPvI0Sy5aKYZJCsKDjHTUUk=";
+        };
+        patches = [ ./0000-seabios-enable-ATA_DMA-4.18.patch ];
+        meta.description = "Xen's fork of SeaBIOS";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalOVMF {
+      "firmware/ovmf-dir-remote" = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/ovmf.git";
+          rev = "edk2-stable${ovmfVersion}";
+          hash = "sha256-htOvV43Hw5K05g0SF3po69HncLyma3BtgpqYSdzRG4s=";
+        };
+        buildInputs = [ nasm ];
+        postPatch = ''
+          substituteInPlace \
+            OvmfPkg/build.sh BaseTools/BinWrappers/PosixLike/{AmlToC,build,GenFfs,GenFv,GenFw,GenSec,LzmaCompress,TianoCompress,Trim,VfrCompile} \
+          --replace-fail "/usr/bin/env bash" ${bash}/bin/bash
+        '';
+        meta.description = "Xen's fork of OVMF";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalIPXE {
+      "firmware/etherboot/ipxe.git" = {
+        src = fetchFromGitHub {
+          owner = "ipxe";
+          repo = "ipxe";
+          rev = "b66e27d9b29a172a097c737ab4d378d60fe01b05";
+          hash = "sha256-TKZ4WjNV2oZIYNefch7E7m1JpeoC/d7O1kofoNv8G40=";
+        };
+        meta.description = "Xen's fork of iPXE";
```

Comment:
It's not a fork, as highlighted best that you're referencing upstream ipxe on github.

Xen has local mirrors of ipxe/seabios/ovmf and the internals of the Xen build refer to a known good configuration, but they are just mirrors.

Reply:
Renamed and pinned to the `rev` in the Makefile.